source/subscription-manager/subscription_manager.c

```diff
@@ -0,0 +1,168 @@
+/*
+ * FreeRTOS V202011.00
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * https://www.FreeRTOS.org
+ * https://aws.amazon.com/freertos
+ *
+ */
+
+/**
+ * @file subscription_manager.c
+ * @brief Functions for managing MQTT subscriptions.
+ */
+
+/* Demo config include. */
+#include "demo_config.h"
+
+/* Subscription manager header include. */
+#include "subscription_manager.h"
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Maximum number of subscriptions maintained by the subscription manager
+ * simultaneously.
+ */
+#ifndef SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS
+    #define SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS    ( 10U )
+#endif
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief An element in the list of subscriptions maintained by the agent.
+ *
+ * @note The agent allows multiple tasks to subscribe to the same topic.
+ * In this case, another element is added to the subscription list, differing
+ * in the intended publish callback.
+ */
+typedef struct subscriptionElement
+{
+    IncomingPubCallback_t pxIncomingPublishCallback;
+    void * pvIncomingPublishCallbackContext;
+    uint16_t uFilterStringLength;
+    const char * pcSubscriptionFilterString;
+} SubscriptionElement_t;
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief The static array of subscription elements.
+ *
+ * @note No thread safety is required to this array, since the updates the array
+ * elements are done only from one task at a time.
+ */
+static SubscriptionElement_t xSubscriptionList[ SUBSCRIPTION_MANAGER_MAX_SUBSCRIPTIONS ];
```

Comment:
I think it would make more sense to use a different list for every MQTT connection. This array can be cast to a void * and passed as the `pIncomingPacketContext` parameter of `MQTTAgent_Init()`, and added as a field to each definition of `CommandContext`